lib/AST/ASTContext.cpp

```diff
@@ -434,10 +434,31 @@ static ModuleDecl *createBuiltinModule(ASTContext &ctx) {
   return M;
 }
 
+inline ASTContext::Implementation &ASTContext::getImpl() const {
+  auto end = reinterpret_cast<void*>(const_cast<ASTContext*>(this) + 1);
+  auto impl = llvm::alignAddr(end, alignof(Implementation));
+  return *reinterpret_cast<Implementation*>(impl);
+}
+
+ASTContext *ASTContext::get(LangOptions &langOpts,
+                            SearchPathOptions &SearchPathOpts,
+                            SourceManager &SourceMgr,
+                            DiagnosticEngine &Diags) {
+  void *mem = nullptr;
+  auto size = sizeof(ASTContext) + sizeof(Implementation);
+  auto alignment = std::max(alignof(ASTContext), alignof(Implementation));
+  int r = posix_memalign(&mem, alignment, size);
```

Comment:
Sorry, I meant that there might be padding between ASTContext and Implementation, requiring a larger size than just the sum.

Reply:
I don't know what the C/C++ spec says about allocators, but I have a hard time imagining any real world allocator where the above code won't work. For example, if code allocates 8 bytes with 16-byte alignment, then what is the disposition of the 8 pad bytes? Who "owns" them? The allocator? Or the caller? Nevertheless, would you like me to handle this scenario by rounding up the size?

Reply:
That's still not the problem. Imagine that ASTContext was 12 bytes long and had 4-byte alignment, and Implementation was 8 bytes long and had 8-byte alignment. Even if you make the initial allocation on an 8-byte alignment, the address `mem + sizeof(ASTContext)` wouldn't be correctly aligned for Implementation.

I don't think this'll come up in practice, but it would be nice™ for the code to be correct. If you don't want to deal with it, I believe `static_assert(alignof(Implementation) <= alignof(ASTContext), "no padding between these types");` would do it. Which is also important for the pointer math in `getImpl()`, I believe.

Reply:
"Allocate 20 bytes with an alignment of 8" does not mean "allocate a multiple of 8 bytes greater than or equal to 20", and ASan will complain at you for assuming that.

Reply:
…oops, but `posix_memalign` does guarantee that. Still, it does still fail if Implementation is also 12 bytes long.

Reply:
\*sigh\* I think you're correct after all: `ASTContext + Implementation + (alignment padding)` = `ASTContext + (alignment padding) + Implementation`. Mind putting in a comment for the next person to have this concern?

Reply:
Just a note: `posix_memalign` isn’t available on all platforms. Something like `AlignedAlloc` in Basic/Malloc.h might be a better choice.